readthedocs/search/serializers.py

```diff
@@ -39,24 +41,59 @@ class PageHighlightSerializer(serializers.Serializer):
 
 class PageSearchSerializer(serializers.Serializer):
 
+    """
+    Page serializer.
+
+    If ``projects_data`` is passed into the context, the serializer
+    will try to use that to generate the link before querying the database.
+    It's a dictionary containing the project slug, and its version URL
+    and version's doctype.
+    """
+
     type = serializers.CharField(default='page', source=None, read_only=True)
     project = serializers.CharField()
     version = serializers.CharField()
     title = serializers.CharField()
     link = serializers.SerializerMethodField()
-    highlight = PageHighlightSerializer(source='meta.highlight', default=dict)
+    highlights = PageHighlightSerializer(source='meta.highlight', default=dict)
     blocks = serializers.SerializerMethodField()
 
     def get_link(self, obj):
-        # TODO: optimize this to not query the db for each result.
-        # TODO: return an relative URL when this is called from the indoc search.
+        """
+        Get the page link.
+
+        Try to get the link from the ``project_data`` context,
+        and fallback to get it from the database.
+        If the result is fetched from the database,
+        it's cached into ``project_data``.
+        """
+        # TODO: return a relative URL when this is called from the indoc search.
+
+        # First try to build the URL from the context.
+        project_data = self.context.get('projects_data', {}).get(obj.project)
+        if project_data:
+            docs_url, doctype = project_data
+            path = obj.full_path
+
+            # Generate an appropriate link for the doctypes that use htmldir,
+            # and always end it with / so it goes directly to proxito.
+            if doctype in {SPHINX_HTMLDIR, MKDOCS}:
+                new_path = re.sub('(^|/)index.html$', '/', path)
+                # docs_url already ends with /,
+                # so path doesn't need to start with /.
+                path = new_path.lstrip('/')
+
+            return docs_url + path
+
+        # Fallback to build the URL querying the db.
         project = Project.objects.filter(slug=obj.project).first()
         if project:
-            return resolve(
-                project=project,
-                version_slug=obj.version,
-                filename=obj.full_path,
-            )
+            docs_url = project.get_docs_url(version_slug=obj.version)
+            # cache the project URL
+            projects_data = self.context.setdefault('projects_data', {})
+            projects_data[obj.project] = (docs_url, '')
```

Comment:
I wonder if we should use some kind of data structure to give these attributes names. It would read much clearer than just setting it to 2 random variables here. Perhaps a NamedTuple?